pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -168,29 +368,91 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
 		// Make a deep copy so we don't mutate the shared cache
 		copy, err := scheme.Scheme.DeepCopy(replicas[i])
 		if err != nil {
-			return err
+			return &status, err
 		}
 		replica := copy.(*v1.Pod)
-		if err := ssc.podControl.UpdateStatefulPod(set, replica); err != nil {
-			return err
+		if err := ssc.podControl.UpdateStatefulPod(updateSet, replica); err != nil {
+			return &status, err
 		}
 	}
 
 	// At this point, all of the current Replicas are Running and Ready, we can consider termination.
 	// We will wait for all predecessors to be Running and Ready prior to attempting a deletion.
 	// We will terminate Pods in a monotonically decreasing order over [len(pods),set.Spec.Replicas).
-	// Note that we do not resurrect Pods in this interval.
+	// Note that we do not resurrect Pods in this interval. Also not that scaling will take precedence over
+	// updates.
 	if unhealthy > 0 && monotonic {
 		glog.V(2).Infof("StatefulSet %s is waiting on %d Pods", set.Name, unhealthy)
-		return nil
+		return &status, nil
 	}
 	for target := len(condemned) - 1; target >= 0; target-- {
 		glog.V(2).Infof("StatefulSet %s terminating Pod %s", set.Name, condemned[target])
 		if err := ssc.podControl.DeleteStatefulPod(set, condemned[target]); err != nil {
-			return err
+			return &status, err
+		}
+		if getPodRevision(pods[target]) == currentRevision.Name {
+			status.CurrentReplicas--
+		} else if getPodRevision(pods[target]) == updateRevision.Name {
+			status.UpdatedReplicas--
 		}
 		if monotonic {
-			return nil
+			return &status, nil
+		}
+	}
+
+	// for the OnDelete strategy we short circuit. Pods will be updated when they are manually deleted.
+	if set.Spec.UpdateStrategy.Type == apps.OnDeleteStatefulSetStrategyType {
+		return &status, nil
+	}
+
+	// we compute the minimum ordinal of the target sequence for a destructive update based on the strategy.
+	updateMin := 0
+	if set.Spec.UpdateStrategy.Type == apps.PartitionStatefulSetStrategyType {
```

Comment:
The code you point to above decides if a Pod should be deleted. 
1. The method short circuits for OnDeleteStatefulSetStrategy. As in the designm Pods are not terminated with this strategy. Update occurs only when the pod is terminated externally.
1. updateMin is set to 0 by default, we can't have an OnDeleteStatefulSetStrategy here because we've returned control above if did.
1. If the strategy is PartitionedStatefulSetStrategy than the updateMin is set to the Partition, otherwise it's still 0, and we have a RollingUpdateStatefulSetStrategy.
1. In the loop below, we walk backward across the replicas to determine the first one that does not match the update revision. We terminate that replica, and we wait for it to be completely deleted before we attempt to apply another update.
1. Therefore, if we are partitioned, we only considered the update partition sequence for targeted termination, and if we are rolling, we consider the entire sequence. For on delete, we do nothing.
1. Also note that, as per the design, scaling takes precedence and is handled prior to termination for update.

The code below creates new replicas
```golang
replicas[i] = newVersionedStatefulSetPod(currentSet,updateSet,currentRevision.Name,updateRevision.Nam,i)
```

If you look at the internals of that method, you will see that it decides which version a replica should be at based on whether it is being tracked as a current or updated replica and on the selected strategy.

```golang
// newVersionedStatefulSetPod creates a new Pod for a StatefulSet. currentSet is the representation of the set at the
// current revision. updateSet is the representation of the set at the updateRevision. currentRevision is the name of
// the current revision. updateRevision is the name of the update revision. ordinal is the ordinal of the Pod. If the
// returned error is nil, the returned Pod is valid.
func newVersionedStatefulSetPod(currentSet *apps.StatefulSet, updateSet *apps.StatefulSet, currentRevision string, updateRevision string, ordinal int) *v1.Pod {
	if (currentSet.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
		ordinal < int(currentSet.Status.CurrentReplicas)) ||
		(currentSet.Spec.UpdateStrategy.Type == apps.PartitionStatefulSetStrategyType &&
			ordinal < int(currentSet.Spec.UpdateStrategy.Partition.Ordinal)) {
		pod := newStatefulSetPod(currentSet, ordinal)
		setPodRevision(pod, currentRevision)
		return pod
	}
	pod := newStatefulSetPod(updateSet, ordinal)
	setPodRevision(pod, updateRevision)
	return pod
}
```
1. If its a current replica we create the replica from the current revision, respecting the partition for the PartitionedStatefulSetStrategy.
1. Otherwise we create it from the update version.